README.md
# Oga

**NOTE:** my spare time is limited which means I am unable to dedicate a lot of
time on Oga. If you're interested in contributing to FOSS, please take a look at
the open issues and submit a pull request to address them where possible.

Oga is an XML/HTML parser written in Ruby. It provides an easy to use API for
parsing, modifying and querying documents (using XPath expressions). Oga does
not require system libraries such as libxml, making it easier and faster to
install on various platforms. To achieve better performance Oga uses a small,
native extension (C for MRI/Rubinius, Java for JRuby).

Oga provides an API that allows you to safely parse and query documents in a
multi-threaded environment, without having to worry about your applications
blowing up.

From [Wikipedia][oga-wikipedia]:

> Oga: A large two-person saw used for ripping large boards in the days before
> power saws. One person stood on a raised platform, with the board below him,
> and the other person stood underneath them.

The name is a pun on [Nokogiri][nokogiri].

## Versioning Policy

Oga uses the version format `MAJOR.MINOR` (e.g. `2.1`). An increase of the MAJOR
version indicates backwards incompatible changes were introduced. The MINOR
version is _only_ increased when changes are backwards compatible, regardless of
whether those changes are bugfixes or new features. Up until version 1.0 the
code should be considered unstable meaning it can change (and break) at any
given moment.

APIs explicitly tagged as private (e.g. using Ruby's `private` keyword or YARD's
`@api private` tag) are not covered by these rules.

## Examples

Parsing a simple string of XML:

    Oga.parse_xml('<people><person>Alice</person></people>')

Parsing XML using strict mode (disables automatic tag insertion):

    Oga.parse_xml('<people>foo</people>', :strict => true) # works fine
    Oga.parse_xml('<people>foo', :strict => true)          # throws an error

Parsing a simple string of HTML:

    Oga.parse_html('<link rel="stylesheet" href="foo.css">')

Parsing an IO handle pointing to XML (this also works when using
`Oga.parse_html`):

    handle = File.open('path/to/file.xml')

    Oga.parse_xml(handle)

Parsing an IO handle using the pull parser:

    handle = File.open('path/to/file.xml')
    parser = Oga::XML::PullParser.new(handle)

    parser.parse do |node|
      parser.on(:text) do
        puts node.text
      end
    end

Using an Enumerator to download and parse an XML document on the fly:

    enum = Enumerator.new do |yielder|
      HTTPClient.get('http://some-website.com/some-big-file.xml') do |chunk|
        yielder << chunk
      end
    end

    document = Oga.parse_xml(enum)

Parse a string of XML using the SAX parser:

    class ElementNames
      attr_reader :names

      def initialize
        @names = []
      end

      def on_element(namespace, name, attrs = {})
        @names << name
      end
    end

    handler = ElementNames.new

    Oga.sax_parse_xml(handler, '<foo><bar></bar></foo>')

    handler.names # => ["foo", "bar"]

Querying a document using XPath:

    document = Oga.parse_xml <<-EOF
    <people>
      <person id="1">
        <name>Alice</name>
        <age>28</name>
      </person>
    </people>
    EOF

    # The "xpath" method returns an enumerable (Oga::XML::NodeSet) that you can
    # iterate over.
    document.xpath('people/person').each do |person|
      puts person.get('id') # => "1"

      # The "at_xpath" method returns a single node from a set, it's the same as
      # person.xpath('name').first.
      puts person.at_xpath('name').text # => "Alice"
    end

Querying the same document using CSS:

    document = Oga.parse_xml <<-EOF
    <people>
      <person id="1">
        <name>Alice</name>
        <age>28</name>
      </person>
    </people>
    EOF

    # The "css" method returns an enumerable (Oga::XML::NodeSet) that you can
    # iterate over.
    document.css('people person').each do |person|
      puts person.get('id') # => "1"

      # The "at_css" method returns a single node from a set, it's the same as
      # person.css('name').first.
      puts person.at_css('name').text # => "Alice"
    end

Modifying a document and serializing it back to XML:

    document = Oga.parse_xml('<people><person>Alice</person></people>')
    name     = document.at_xpath('people/person[1]/text()')

    name.text = 'Bob'

    document.to_xml # => "<people><person>Bob</person></people>"

Querying a document using a namespace:

    document = Oga.parse_xml('<root xmlns:x="foo"><x:div></x:div></root>')
    div      = document.xpath('root/x:div').first

    div.namespace # => Namespace(name: "x" uri: "foo")

## Features

* Support for parsing XML and HTML(5)
  * DOM parsing
  * Stream/pull parsing
  * SAX parsing
* Low memory footprint
* High performance (taking into account most work happens in Ruby)
* Support for XPath 1.0
* CSS3 selector support
* XML namespace support (registering, querying, etc)
* Windows support

## Requirements

| Ruby     | Required      | Recommended |
|:---------|:--------------|:------------|
| MRI      | >= 2.3.0      | >= 2.6.0    |
| JRuby    | >= 1.7        | >= 1.7.12   |
| Rubinius | Not supported |             |
| Maglev   | Not supported |             |
| Topaz    | Not supported |             |
| mruby    | Not supported |             |

Maglev and Topaz are not supported due to the lack of a C API (that I know of)
and the lack of active development of these Ruby implementations. mruby is not
supported because it's a very different implementation all together.

To install Oga on MRI or Rubinius you'll need to have a working compiler such as
gcc or clang. Oga's C extension can be compiled with both. JRuby does not
require a compiler as the native extension is compiled during the Gem building
process and bundled inside the Gem itself.

## Thread Safety

Oga does not use a unsynchronized global mutable state. As a result of this you
can parse/create documents concurrently without any problems. Modifying
documents concurrently can lead to bugs as these operations are not
synchronized.

Some querying operations will cache data in instance variables, without
synchronization. An example is `Oga::XML::Element#namespace` which will cache an
element's namespace after the first call.

In general it's recommended to _not_ use the same document in multiple threads
at the same time.

## Namespace Support

Oga fully supports parsing/registering XML namespaces as well as querying them
using XPath. For example, take the following XML:

    <root xmlns="http://example.com">
        <bar>bar</bar>
    </root>

If one were to try and query the `bar` element (e.g. using XPath `root/bar`)
they'd end up with an empty node set. This is due to `<root>` defining an
alternative default namespace. Instead you can query this element using the
following XPath:

    *[local-name() = "root"]/*[local-name() = "bar"]

Alternatively, if you don't really care where the `<bar>` element is located you
can use the following:

    descendant::*[local-name() = "bar"]

And if you want to specify an explicit namespace URI, you can use this:

    descendant::*[local-name() = "bar" and namespace-uri() = "http://example.com"]

Unlike Nokogiri, Oga does _not_ provide a way to create "dynamic" namespaces.
That is, Nokogiri allows one to query the above document as following:

    document = Nokogiri::XML('<root xmlns="http://example.com"><bar>bar</bar></root>')

    document.xpath('x:root/x:bar', :x => 'http://example.com')

Oga does have a small trick you can use to cut down the size of your XPath
queries. Because Oga assigns the name "xmlns" to default namespaces you can use
this in your XPath queries:

    document = Oga.parse_xml('<root xmlns="http://example.com"><bar>bar</bar></root>')

    document.xpath('xmlns:root/xmlns:bar')

When using this you can still restrict the query to the correct namespace URI:

    document.xpath('xmlns:root[namespace-uri() = "http://example.com"]/xmlns:bar')

In the future I might add an API to ease this process, although at this time I
have little interest in providing an API similar to Nokogiri.

## HTML5 Support

Oga fully supports HTML5 including the omission of certain tags. For example,
the following is parsed just fine:

    <li>Hello
    <li>World

This is effectively parsed into:

    <li>Hello</li>
    <li>World</li>

One exception Oga makes is that it does _not_ automatically insert `html`,
`head` and `body` tags. Automatically inserting these tags requires a
distinction between documents and fragments as a user might not always want
these tags to be inserted if left out. This complicates the user facing API as
well as complicating the parsing internals of Oga. As a result I have decided
that Oga _does not_ insert these tags when left out.

A more in depth explanation can be found here:
<https://gitlab.com/yorickpeterse/oga/issues/98#note_45443992>

## Documentation

The documentation is best viewed [on the documentation website][doc-website].

* {file:CONTRIBUTING Contributing}
* {file:changelog Changelog}
* {file:migrating\_from\_nokogiri Migrating From Nokogiri}
* {Oga::XML::Parser XML Parser}
* {Oga::XML::SaxParser XML SAX Parser}
* {file:xml\_namespaces XML Namespaces}

## Why Another HTML/XML parser?

Currently there are a few existing parser out there, the most famous one being
[Nokogiri][nokogiri]. Another parser that's becoming more popular these days is
[Ox][ox]. Ruby's standard library also comes with REXML.

The sad truth is that these existing libraries are problematic in their own
ways. Nokogiri for example is extremely unstable on Rubinius. On MRI it works
because of the non concurrent nature of MRI, on JRuby it works because it's
implemented as Java. Nokogiri also uses libxml2 which is a massive beast of a
library, is not thread-safe and problematic to install on certain platforms
(apparently). I don't want to compile libxml2 every time I install Nokogiri
either.

To give an example about the issues with Nokogiri on Rubinius (or any other
Ruby implementation that is not MRI or JRuby), take a look at these issues:

* <https://github.com/rubinius/rubinius/issues/2957>
* <https://github.com/rubinius/rubinius/issues/2908>
* <https://github.com/rubinius/rubinius/issues/2462>
* <https://github.com/sparklemotion/nokogiri/issues/1047>
* <https://github.com/sparklemotion/nokogiri/issues/939>

Some of these have been fixed, some have not. The core problem remains:
Nokogiri acts in a way that there can be a large number of places where it
*might* break due to throwing around void pointers and what not and expecting
that things magically work. Note that I have nothing against the people running
these projects, I just heavily, *heavily* dislike the resulting codebase one
has to deal with today.

Ox looks very promising but it lacks a rather crucial feature: parsing HTML
(without using a SAX API). It's also again a C extension making debugging more
of a pain (at least for me).

I just want an XML/HTML parser that I can rely on stability wise and that is
written in Ruby so I can actually debug it. In theory it should also make it
easier for other Ruby developers to contribute.

## License

All source code in this repository is subject to the terms of the Mozilla Public
License, version 2.0 unless stated otherwise. A copy of this license can be
found the file "LICENSE" or at <https://www.mozilla.org/MPL/2.0/>.

[nokogiri]: https://github.com/sparklemotion/nokogiri
[oga-wikipedia]: https://en.wikipedia.org/wiki/Japanese_saw#Other_Japanese_saws
[ox]: https://github.com/ohler55/ox
[doc-website]: http://code.yorickpeterse.com/oga/latest/


CHANGELOG.md
# Changelog

This document contains details of the various releases and their release dates.
Dates are in the format `yyyy-mm-dd`.

## 3.4 - 2022-08-02

This release includes a change that when setting the child nodes of a node A,
node A takes ownership over the entire new child tree. See merge request
https://gitlab.com/yorickpeterse/oga/-/merge_requests/194 for more details.

## 3.3 - 2020-07-27

This release adds `to_s` as an alias for `to_xml`, thanks to Roy Zwambag. See
merge request https://gitlab.com/yorickpeterse/oga/-/merge_requests/192 for more
information.

## 3.2 - 2020-01-10

This release fixes a few warnings that would show up when using Oga on Ruby
2.7.0. See https://gitlab.com/yorickpeterse/oga/merge_requests/190 for more
information.

## 3.1 - 2020-01-08

This release fixes a bug in the XML lexer that prevented the parsing of doctypes
using "public" or "system" instead of "PUBLIC"/"SYSTEM". See issue
<https://gitlab.com/yorickpeterse/oga/issues/199> for more information.

## 3.0 - 2019-12-03

This release bumps the Ruby version requirement to Ruby 2.3.0, as we haven't
supported older versions for several years now. We also no longer officially
support Rubinius.

## 2.17 - 2019-12-02

Elements using the default XML namespace can now be queried using XPath queries,
which was broken for quite a while.

See commit <https://gitlab.com/yorickpeterse/oga/commit/95da93949bf613612981f5cd7decc0d2c2a60e15>
for more information.

## 2.16 - 2019-11-29

* XPath namespace aliases can now be used when querying elements using XPath
  expressions.
* Several RDOc and RubyGems deprecation warnings have been resolved.

See the following commits for more information:

* <https://gitlab.com/yorickpeterse/oga/commit/d9e7346b60c3afa2b3e83a240f9807c6bb819d48>
* <https://gitlab.com/yorickpeterse/oga/commit/da9721cb34f91527e72b096c6bd6a128e37b1992>
* <https://gitlab.com/yorickpeterse/oga/commit/977bd594c8bfd1a29aeba9d3a4ab7d0ebbc7d11a>

## 2.15 - 2018-04-11

The HTML parser now allows `th` elements to occur in `thead`, `tbody`, and
`tfoot` elements. See issue <https://gitlab.com/yorickpeterse/oga/issues/190>
for more information.

## 2.14 - 2018-01-30

Various methods that yield a block now return an Enumerator when no block is
provided. See merge request
<https://gitlab.com/yorickpeterse/oga/merge_requests/184> for more information.

## 2.13 - 2018-01-05

Leading and trailing whitespace is now removed from CSS selectors. See
<https://gitlab.com/yorickpeterse/oga/merge_requests/183> for more information.

## 2.12 - 2017-12-29

Element start tags containing other start tags (e.g. `<script<script>`) are now
parsed correctly.

See f574197ea657cf09405336ca618a22e32c94d0d0 for more information.

## 2.11 - 2017-09-07

Various Ruby warnings have been resolved by Loic Nageleisen. See pull request
<https://gitlab.com/yorickpeterse/oga/pull/180> for more information.

## 2.10 - 2017-04-18

### Fix `Element#attribute` for HTML documents when using Symbol arguments

You can now pass a Symbol to `Oga::XML::Element#attribute` for both XML and HTML
documents, previously this only worked for XML documents. See
[PR #174](https://gitlab.com/yorickpeterse/oga/pull/174) for more information.

## 2.9 - 2017-02-10

### Closing tags for HTML void elements

Certain HTML elements such as `<img>` and `<link>` (called "void elements" in
Oga) are now closed using a `>` tag instead of `/>`. In other words, instead of
outputting `<img src="..." />` Oga now outputs `<img src="...">`.

### Doctypes are now Nodes

Each Doctype now inherits from `Oga::XML::Node`. This makes it possible to parse
documents where a doctype is located in a child node. However, in these cases
Oga will _not_ populate `Oga::XML::Document#doctype` as this can not be done in
an efficient way.

## 2.8 - 2017-01-04

Ruby 2.4 deprecates Fixnum in favour of Integer, producing warnings whenever
Fixnum is used. Oga 2.8 contains a fix contributed by Po Shan Cheah to remove
these deprecation warnings. See commit c75ca96d229a50b369e16057622255a674f2cabc
for more information.

## 2.7 - 2016-09-27

### Closing Elements When Generating XML

When generating XML Oga now properly closes elements with siblings but without
children. See commit e0e0687dc29427c854c9fa6d3c19cee1c04f92c7 for more
information.

### Newlines After Doctypes

When generating XML a newline would be inserted after a doctype. If another
newline would follow in a text node this would lead to multiple newlines being
present. Oga now ensures there is only 1 newline following a doctype. See commit
e0e0687dc29427c854c9fa6d3c19cee1c04f92c7 for more information.

### Processing Instructions With Namespace Prefixes

The XML lexer now supports processing instructions containing namespace prefixes
such as `<?xml:foo ?>`. See commit 01fa1513f4bd6f194bf6d1ca17e510003fa23312 for
more information.

### XML Declarations Are Now Processing Instructions

The class `Oga::XML::XmlDeclaration` now extends
`Oga::XML::ProcessingInstruction`. This allows documents to contain XML
declarations in nested elements, instead of only allowing this at the root of
the document. See commit 116b9b0ceb6feab4daa0bb417302590fba948bef for more
information.

### Aliases For Getting & Setting Attributes

The methods `Oga::XML::Element#get` and `Oga::XML::Element#set` are now aliased
as `#[]` and `#[]=` respectively. See d40baf0c724a3874f43100fbefa775cfb8dcacda
for more information and thanks to Scott Wheeler for contributing the patch.

## 2.6 - 2016-09-10

This release fixes a bug in the XML generation code that would cause it to get
stuck in the generation loop. See issue
<https://gitlab.com/yorickpeterse/oga/issues/161> and commit
38284278d542640c3d8300ef15890af93b6df779 for more information.

## 2.5 - 2016-09-06

This release fixes a bug in the XML parser that would prevent it from parsing
doctypes that contain a mixture of public/system IDs, a name, and inline rules.


lib/oga/parser.rb
#
# DO NOT MODIFY!!!!
# This file is automatically generated by Racc 1.4.11
# from Racc grammer file "".
#

require 'racc/parser.rb'
module Oga
  class Parser < Racc::Parser


  def initialize
    @lexer = Lexer.new
  end

  def reset
    @lines  = []
    @line   = 1
    @column = 1
  end

  def s(type, *children)
    return AST::Node.new(
      type,
      children.flatten,
      :line   => @line,
      :column => @column
    )
  end

  def next_token
    type, value, line, column = @tokens.shift

    @line   = line if line
    @column = column if column

    return type ? [type, value] : [false, false]
  end

  def on_error(type, value, stack)
    name      = token_to_str(type)
    line_str  = @lines[@line - 1]
    indicator = '~' * (@column - 1) + '^'

    raise Racc::ParseError, <<-EOF.strip
Failed to parse the supplied input.

Reason:   unexpected #{name} with value #{value.inspect}
Location: line #{@line}, column #{@column}

Offending code:

#{line_str}
#{indicator}

Current stack:

#{stack.inspect}
    EOF
  end

  def parse(string)
    @lines  = string.lines
    @tokens = @lexer.lex(string)
    ast     = do_parse

    reset

    return ast
  end

# vim: set ft=racc:
##### State transition tables begin ###

racc_action_table = [
     7,     7,    14,    11,    13,     9,    15,    16,    18,    19,
    20,   nil,     6,     6 ]

racc_action_check = [
     0,     2,     9,     7,     8,     1,    11,    12,    16,    18,
    19,   nil,     0,     2 ]

racc_action_pointer = [
    -4,     5,    -3,   nil,   nil,   nil,   nil,   -12,   -11,     2,
   nil,     1,     3,   nil,   nil,   nil,     2,   nil,    -6,     5,
   nil ]

racc_action_default = [
    -2,   -16,    -1,    -4,    -5,    -6,    -7,   -16,   -16,   -16,
    -3,   -16,   -16,   -11,    21,    -8,   -16,   -10,   -16,   -16,
    -9 ]

racc_goto_table = [
     3,     1,    10,     2,    17,    12 ]

racc_goto_check = [
     3,     1,     3,     2,     7,     8 ]

racc_goto_pointer = [
   nil,     1,     3,     0,   nil,   nil,   nil,    -8,    -3,   nil,
   nil ]

racc_goto_default = [
   nil,   nil,   nil,   nil,     4,     5,     8,   nil,   nil,   nil,
   nil ]

racc_reduce_table = [
  0, 0, :racc_error,
  1, 18, :_reduce_1,
  0, 18, :_reduce_2,
  2, 19, :_reduce_3,
  1, 19, :_reduce_4,
  1, 20, :_reduce_none,
  1, 20, :_reduce_none,
  1, 22, :_reduce_7,
  3, 23, :_reduce_8,
  4, 24, :_reduce_none,
  3, 21, :_reduce_10,
  1, 25, :_reduce_none,
  2, 26, :_reduce_none,
  1, 26, :_reduce_none,
  1, 27, :_reduce_none,
  1, 27, :_reduce_none ]

racc_reduce_n = 16

racc_shift_n = 21

racc_token_table = {
  false => 0,
  :error => 1,
  :T_SPACE => 2,
  :T_NEWLINE => 3,
  :T_SMALLER => 4,
  :T_GREATER => 5,
  :T_SLASH => 6,
  :T_DQUOTE => 7,
  :T_SQUOTE => 8,
  :T_DASH => 9,
  :T_RBRACKET => 10,
  :T_LBRACKET => 11,
  :T_COLON => 12,
  :T_BANG => 13,
  :T_EQUALS => 14,
  :T_TEXT => 15,
  :T_DOCTYPE => 16 }

racc_nt_base = 17

racc_use_result_var = false

Racc_arg = [
  racc_action_table,
  racc_action_check,
  racc_action_default,
  racc_action_pointer,
  racc_goto_table,
  racc_goto_check,
  racc_goto_default,
  racc_goto_pointer,
  racc_nt_base,
  racc_reduce_table,
  racc_token_table,
  racc_shift_n,
  racc_reduce_n,
  racc_use_result_var ]

Racc_token_to_s_table = [
  "$end",
  "error",
  "T_SPACE",
  "T_NEWLINE",
  "T_SMALLER",
  "T_GREATER",
  "T_SLASH",
  "T_DQUOTE",
  "T_SQUOTE",
  "T_DASH",
  "T_RBRACKET",
  "T_LBRACKET",
  "T_COLON",
  "T_BANG",
  "T_EQUALS",
  "T_TEXT",
  "T_DOCTYPE",
  "$start",
  "document",
  "expressions",
  "expression",
  "tag",
  "doctype",
  "tag_start",
  "tag_end",
  "tag_body",
  "whitespaces",
  "whitespace" ]

Racc_debug_parser = false

##### State transition tables end #####

# reduce 0 omitted

def _reduce_1(val, _values)
 s(:document, val[0]) 
end

def _reduce_2(val, _values)
 s(:document) 
end

def _reduce_3(val, _values)
 val.compact 
end

def _reduce_4(val, _values)
 val[0] 
end

# reduce 5 omitted

# reduce 6 omitted

def _reduce_7(val, _values)
 s(:doctype, val[0]) 
end

def _reduce_8(val, _values)
 val[1] 
end

# reduce 9 omitted

def _reduce_10(val, _values)
        s(:element, val[0], val[1])
      
end

# reduce 11 omitted

# reduce 12 omitted

# reduce 13 omitted

# reduce 14 omitted

# reduce 15 omitted

def _reduce_none(val, _values)
  val[0]
end

  end   # class Parser
  end   # module Oga
